src/rust/engine/process_execution/src/nailgun/parsed_jvm_command_lines.rs

```diff
@@ -0,0 +1,317 @@
+use itertools::Itertools;
+use std::slice::Iter;
+
+/// Represents the result of parsing the args of a nailgunnable ExecuteProcessRequest
+/// TODO(8481) We may want to split the classpath by the ":", and store it as a Vec<String>
+///         to allow for deep fingerprinting.
+#[derive(PartialEq, Eq, Debug)]
+pub struct ParsedJVMCommandLines {
+    pub nailgun_args: Vec<String>,
+    pub client_args: Vec<String>,
+    pub client_main_class: String,
+}
+
+impl ParsedJVMCommandLines {
+    ///
+    /// Given a list of args that one would likely pass to a java call,
+    /// we automatically split it to generate two argument lists:
+    ///  - nailgun arguments: The list of arguments needed to start the nailgun server.
+    ///    These arguments include everything in the arg list up to (but not including) the main class.
+    ///    These arguments represent roughly JVM options (-Xmx...), and the classpath (-cp ...).
+    ///
+    ///  - client arguments: The list of arguments that will be used to run the jvm program under nailgun.
+    ///    These arguments can be thought of as "passthrough args" that are sent to the jvm via the nailgun client.
+    ///    These arguments include everything starting from the main class.
+    ///
+    /// We assume that:
+    ///  - Every args list has a main class.
+    ///  - There is exactly one argument that doesn't begin with a `-` in the command line before the main class,
+    ///    and it's the value of the classpath (i.e. `-cp scala-library.jar`).
+    ///
+    /// We think these assumptions are valid as per: https://github.com/pantsbuild/pants/issues/8387
```

Comment:
I considered it briefly, here is my thought process:
- We could use a CLI arg parsing library, like `clap`. However, that would involve re-implementing the JVM CLI specification in that library. Most of these libs are opinionated (like `scopt` with one-dash options), so that might be more trouble than it's worth.

- We _could_ use a parsing library, as we really just want to parse the context-free grammar:
```
CLI => jvm_args* classpath jvm_args* main_class client_args
jvm_args => r"-[^\w]+"
classpath => ("-cp"|"-classpath") "[^-][^\w*]"
main_class => "[^-][^\w*]"
client_args => "[@,-][^\w]+"
``` 
This looks like any LR1 parser could handle it. However, when I looked into parsing libs for Rust, the options were not very compelling: Either use an antlr-like "your grammar is defined in a DSL in a separate file" lib, or `nom`, which is a great parser combinator lib, but it's geared towards byte-level operations.
In both cases, it seemed easier to implement the parsing by hand for our case.

However, I might have missed something, definitely open to suggestions. In particular, if there is any lib that implements the spec for JVM CLIs.

Reply:
We should likely strongly consider this if we learn that we need to go beyond parsing of the `-cp` arg.

Reply:
If we need to go beyond this, I think I'd like to revisit the UX for rule writers (explicit specification of nailgun vs implicit parsing).